From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: way-zer <himc.wicp@gmail.com>
Date: Sun, 28 Jan 2024 17:39:57 +0800
Subject: [PATCH] =?UTF-8?q?HC:=20Loader=20Mod=E5=AE=9E=E7=8E=B0?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

添加`icon.png`
way-zer <himc.wicp@gmail.com> on 2024/6/9

更好的Desktop适配
way-zer <himc.wicp@gmail.com> on 2024/8/3
---
 .../src/mindustryX/loader/AndroidImpl.java    | 328 ++++++++++++++++++
 core/src/mindustry/mod/ModClassLoader.java    |   7 +
 2 files changed, 335 insertions(+)
 create mode 100644 android/src/mindustryX/loader/AndroidImpl.java

diff --git a/android/src/mindustryX/loader/AndroidImpl.java b/android/src/mindustryX/loader/AndroidImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..6e87ee9f906947eb937b9e3eaf371784460e9e2c
--- /dev/null
+++ b/android/src/mindustryX/loader/AndroidImpl.java
@@ -0,0 +1,328 @@
+package mindustryX.loader;
+
+import android.app.*;
+import android.content.pm.*;
+import android.os.*;
+import android.view.*;
+import arc.*;
+import arc.Files.*;
+import arc.backend.android.*;
+import arc.files.*;
+import arc.func.*;
+import arc.util.*;
+import dalvik.system.*;
+import mindustry.*;
+import mindustry.android.*;
+import mindustry.ui.FileChooser.*;
+
+import java.io.*;
+import java.lang.reflect.Proxy;
+import java.lang.reflect.*;
+import java.net.*;
+import java.util.*;
+
+public class AndroidImpl implements LoaderPlatform{
+    private AndroidLauncher app = (AndroidLauncher)Core.app;
+
+    private Object[] getDexElements(ClassLoader classLoader){
+        Object pathList = Reflect.get(BaseDexClassLoader.class, classLoader, "pathList");
+        return Reflect.get(pathList, "dexElements");
+    }
+
+    private File getElementPath(Object element){
+        try{
+            return Reflect.get(element, "path");
+        }catch(Throwable ignored){
+            return Reflect.get(element, "file");
+        }
+    }
+
+    private File findFirstJar(){
+        return getElementPath(getDexElements(Main.class.getClassLoader())[0]);
+    }
+
+    @Override
+    public void withSafeClassloader(String method){
+        try{
+            ClassLoader classLoader = new DexClassLoader(findFirstJar().getPath(), app.getFilesDir().getPath(), app.getApplicationInfo().nativeLibraryDir, Core.class.getClassLoader());
+            Reflect.invoke(classLoader.loadClass(Main.class.getName()), method);
+        }catch(Exception e){
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public ClassLoader createClassloader(){
+        ClassLoader parent = Core.class.getClassLoader();
+        File apk = getElementPath(getDexElements(parent)[0]);
+        ClassLoader classLoader = new DexClassLoader(findFirstJar().getPath() + File.pathSeparator + apk.getPath(), app.getFilesDir().getPath(), app.getApplicationInfo().nativeLibraryDir, parent){
+            @Override
+            protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException{
+                if(!overwrite(name))
+                    return super.loadClass(name, resolve);
+                //check for loaded state
+                Class<?> loadedClass = findLoadedClass(name);
+                if(loadedClass == null){
+                    try{
+                        //try to load own class first
+                        loadedClass = findClass(name);
+                    }catch(ClassNotFoundException e){
+                        //use parent if not found
+                        return parent.loadClass(name);
+                    }
+                }
+
+                return loadedClass;
+            }
+
+            private Boolean overwrite(String name){
+//                if(name.contains("ExternalSynthetic")) return false;
+                if(name.startsWith("arc.func")) return false;
+                return name.startsWith("mindustry") || name.startsWith("arc.") || name.startsWith("rhino.");
+            }
+
+            @Override
+            public String findLibrary(String name){
+                //TODO Soloud的native会查找class，而且不支持classloader。
+                String file = super.findLibrary(name);
+                //Android的jni方法不能跨classloader，且同一个so库加载多次
+                if(file != null){
+                    Fi newFile = new Fi(app.getFilesDir()).child(name + ".so");
+                    new Fi(file).copyTo(newFile);
+                    file = newFile.path();
+                }
+                return file;
+            }
+
+            @Override
+            public URL getResource(String name){
+                if(name.equals("version.properties"))
+                    return parent.getResource(name);
+                //self first
+                URL url = findResource(name);
+                if(url == null)
+                    url = parent.getResource(name);
+                return url;
+            }
+
+            @Override
+            public Enumeration<URL> getResources(String name) throws IOException{
+                return new CompoundURLEnumeration(
+                //self first
+                findResources(name), parent.getResources(name)
+                );
+            }
+        };
+        return classLoader;
+    }
+
+    @Override
+    public void launch(ClassLoader loader) throws Exception{
+
+        Class<?> cls = loader.loadClass(AndroidImpl.class.getName());
+        app.handler.post(() -> {
+            ((AndroidInput)Core.input).onPause();
+//            Reflect.invoke(Core.graphics, "pause");
+            Core.graphics.dispose();
+            Core.audio.dispose();
+            app.setContentView(Reflect.invoke(cls, "bootStrapAsView",
+            new Object[]{app, app.getListeners().get(0)}, Activity.class, Object.class));
+        });
+    }
+
+    public static View bootStrapAsView(Activity activity, Object platform){
+        AndroidApplication newApp = new AndroidLauncher();
+        copyFields(activity, newApp);
+        AndroidApplicationConfiguration config = new AndroidApplicationConfiguration();
+        config.useImmersiveMode = true;
+        config.hideStatusBar = true;
+        config.useGL30 = true;
+        ClientLauncher client = new ClientLauncher(){
+
+            @Override
+            public void hide(){
+                activity.moveTaskToBack(true);
+            }
+
+            @Override
+            public rhino.Context getScriptContext(){
+                return AndroidRhinoContext.enter(activity.getCacheDir());
+            }
+
+            @Override
+            public void shareFile(Fi file){
+            }
+
+            @Override
+            public ClassLoader loadJar(Fi jar, ClassLoader parent) throws Exception{
+                //Required to load jar files in Android 14: https://developer.android.com/about/versions/14/behavior-changes-14#safer-dynamic-code-loading
+                jar.file().setReadOnly();
+                return new DexClassLoader(jar.file().getPath(), activity.getFilesDir().getPath(), activity.getApplicationInfo().nativeLibraryDir, parent){
+                    @Override
+                    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException{
+                        //check for loaded state
+                        Class<?> loadedClass = findLoadedClass(name);
+                        if(loadedClass == null){
+                            try{
+                                //try to load own class first
+                                loadedClass = findClass(name);
+                            }catch(ClassNotFoundException | NoClassDefFoundError e){
+                                //use parent if not found
+                                return parent.loadClass(name);
+                            }
+                        }
+
+                        if(resolve){
+                            resolveClass(loadedClass);
+                        }
+                        return loadedClass;
+                    }
+                };
+            }
+
+            @Override
+            public void showFileChooser(FileChooserParams params){
+                try{
+                    ClassLoader platformLoader = platform.getClass().getClassLoader();
+                    Class<?> paramsClass = platformLoader.loadClass("mindustry.ui.FileChooser$FileChooserParams");
+                    Class<?> consClass = platformLoader.loadClass("arc.func.Cons");
+                    Object platformParams = paramsClass.getDeclaredConstructor().newInstance();
+
+                    for(Field field : paramsClass.getFields()){
+                        String name = field.getName();
+                        if(name.equals("handler") || name.equals("multipleHandler")) continue;
+                        try{
+                            field.set(platformParams, Reflect.get(params, name));
+                        }catch(Throwable ignored){
+                        }
+                    }
+
+                    if(params.handler != null){
+                        Reflect.set(platformParams, "handler", Proxy.newProxyInstance(platformLoader, new Class[]{consClass}, (proxy, method, args) -> {
+                            if(method.getName().equals("get") && args != null && args.length == 1){
+                                params.handler.get(convertFi(args[0]));
+                            }
+                            return null;
+                        }));
+                    }
+                    if(params.multipleHandler != null){
+                        Reflect.set(platformParams, "multipleHandler", Proxy.newProxyInstance(platformLoader, new Class[]{consClass}, (proxy, method, args) -> {
+                            if(method.getName().equals("get") && args != null && args.length == 1 && args[0] instanceof Object[]){
+                                Object[] source = (Object[])args[0];
+                                Fi[] files = new Fi[source.length];
+                                for(int i = 0; i < source.length; i++) files[i] = convertFi(source[i]);
+                                params.multipleHandler.get(files);
+                            }
+                            return null;
+                        }));
+                    }
+
+                    Reflect.invoke(platformParams, "checkParams");
+                    Reflect.invoke(platform, "showFileChooser", new Object[]{platformParams}, paramsClass);
+                }catch(Throwable e){
+                    Log.err(e);
+                }
+            }
+
+            private Fi convertFi(Object fi){
+                Class<?> cls = fi.getClass();
+                if(cls.isAnonymousClass()){
+                    return new Fi((File)Reflect.get(fi, "file")){
+                        @Override
+                        public InputStream read(){
+                            return Reflect.invoke(fi, "read");
+                        }
+
+                        @Override
+                        public OutputStream write(boolean append){
+                            return Reflect.invoke(fi, "write", new Object[]{append}, boolean.class);
+                        }
+                    };
+                }else if(cls.getSimpleName().equals("Fi")){
+                    return new Fi((File)Reflect.get(fi, "file"));
+                }else{
+                    Vars.ui.showErrorMessage("Not Implement showFileChooser");
+                    return null;
+                }
+            }
+
+            @Override
+            public void beginForceLandscape(){
+                activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_SENSOR_LANDSCAPE);
+            }
+
+            @Override
+            public void endForceLandscape(){
+                activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_USER);
+            }
+        };
+        View view = newApp.initializeForView(client, config);
+        activity.getApplication().registerActivityLifecycleCallbacks(new android.app.Application.ActivityLifecycleCallbacks(){
+            @Override
+            public void onActivityCreated(Activity activity, Bundle savedInstanceState){
+            }
+
+            @Override
+            public void onActivityStarted(Activity activity){
+            }
+
+            @Override
+            public void onActivityResumed(Activity activity){
+                client.resume();
+            }
+
+            @Override
+            public void onActivityPaused(Activity activity){
+                client.pause();
+            }
+
+            @Override
+            public void onActivityStopped(Activity activity){
+            }
+
+            @Override
+            public void onActivitySaveInstanceState(Activity activity, Bundle outState){
+            }
+
+            @Override
+            public void onActivityDestroyed(Activity activity){
+                client.dispose();
+            }
+        });
+        ClassLoader loader = Main.class.getClassLoader();
+        assert loader != null;
+        Files origin = Core.files;
+        Core.files = (Files)Proxy.newProxyInstance(Main.class.getClassLoader(), new Class[]{Files.class}, (proxy, method, args) -> {
+            if(method.getName().equals("internal") || (method.getName().equals("get") && args[1] == FileType.internal)){
+                String path = (String)args[0];
+                if(loader.getResource(path) != null){
+                    return origin.classpath(path);
+                }
+            }
+            return method.invoke(origin, args);
+        });
+        try{
+            //new external folder
+            Fi data = Core.files.absolute(newApp.getExternalFilesDir(null).getAbsolutePath());
+            Core.settings.setDataDirectory(data);
+        }catch(Exception e){
+            //print log but don't crash
+            Log.err(e);
+        }
+        return view;
+    }
+
+    private static void copyFields(Object a, Object b){
+        Class<?> ca = a.getClass(), cb = b.getClass();
+        while(ca != Object.class){
+            if(ca == cb){
+                for(Field f : ca.getDeclaredFields()){
+                    f.setAccessible(true);
+                    Reflect.set(b, f, Reflect.get(a, f));
+                }
+            }
+            ca = ca.getSuperclass();
+            cb = cb.getSuperclass();
+        }
+    }
+}
diff --git a/core/src/mindustry/mod/ModClassLoader.java b/core/src/mindustry/mod/ModClassLoader.java
index d9c05036f9715933ef293511a2069d65312d01fc..8d3a22272e38698dac6d9f0c74ffd4932e3a4fc2 100644
--- a/core/src/mindustry/mod/ModClassLoader.java
+++ b/core/src/mindustry/mod/ModClassLoader.java
@@ -1,6 +1,7 @@
 package mindustry.mod;
 
 import arc.struct.*;
+import java.net.*;
 
 public class ModClassLoader extends ClassLoader{
     private Seq<ClassLoader> children = new Seq<>();
@@ -46,4 +47,10 @@ public class ModClassLoader extends ClassLoader{
 
         throw (last == null ? new ClassNotFoundException(name) : last);
     }
+
+    @Override
+    public URL getResource(String name){
+        if(name.equals("mod.hjson") || name.equals("icon.png")) return null;
+        return super.getResource(name);
+    }
 }
